File: algorithms/CPlusPlus/Strings/anagram.cpp
/* 
   Description : To check whether a string is anagram or not.
   Anagram: An anagram is a word or phrase formed by rearranging the letters in another word or phrase
   Approach:Using characters as array index
*/
#include<bits/stdc++.h>
using namespace std;

bool isAnagram(string s1, string s2) {
    int temp[256] = { 0 };
    for (int i = 0; i < s1.length(); i++) {         //Idea : Here we are using characters of the string as indexes.
        temp[s1[i]]++;                              //Each character(except same) will have different position allocated in the temp array      
        temp[s2[i]]--;                              //Here, characters of first string are incrementing the count at s1[i] position and characters
                                                    //of second string are decrementing the count at s2[i] position
    }
    for (int i = 0; i < 256; i++) {
        if (temp[i] > 0) {                         //If we found any index in temp array with value greater than 0 then the
            return false;                          //strings are not anagram
        }
    }
    return true;                                   //If values at all the indexes of temp array is zero, this will indicate that all characters
}                                                  //in second array cancelled the count of characters in the first array
int main() {
    string s1 ;
    string s2 ;
    cin>>s1;
    cin>>s2;
    string ans = isAnagram(s1, s2) ? "YES" : "NO";
    cout << ans << endl;
    return 0;
    //Ex :  s1= "abc" s2="abd" 
    //Output : NO
    //Time Complexity : O(n), where  n is the length of string
}

File: algorithms/Go/strings/anagram.go
/**
Problem Statement: Given two strings s and t, find they are anagrams or not
An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

Example 1: Input: s="anagram" t="nagaram"
		   Output: true

Example 2: Input: s="rat" t="car"
		   Output: false

Time complexity: O(n), n=length of string s
Space complexity: O(1)

**/

package strings

import "fmt"

func RunIsAnagram() {
	fmt.Println("Enter first string")
	var first string
	fmt.Scanln(&first)
	fmt.Println("Enter second string")
	var second string
	fmt.Scanln(&second)
	ans := isAnagram(first, second)
	if ans == true {
		fmt.Println("They are anagrams")
	} else {
		fmt.Println("They are not anagrams")
	}
}

func isAnagram(s string, t string) bool {
	var count [256]int

	for i := 0; i < len(s); i++ {
		count[s[i]]++
	}
	for i := 0; i < len(t); i++ {
		count[t[i]]--
	}
	for i := 0; i < len(count); i++ {
		if count[i] != 0 {
			return false
		}
	}
	return true
}
